### DISCUSSION TITLE
Promoting `eslint-define-config`?

### DISCUSSION THREAD

**Discussion by Lehoczky** [typescript-eslint/typescript-eslint#8802] [Technical Discussions]
### Body

Hey Guys 👋 

I know I'm late for the party, because `tseslint.config()` was introduced more than a month ago, but I wanted to ask regardless what do you think about using [eslint-define-config ](https://github.com/eslint-types/eslint-define-config) for declaring the ESLint config in a type safe manner?

I feel like @Shinigami92 did a great job with that library and it would be great to join forces with him to have a de-facto solution for type-safe configurations. It feels like we have two competing solutions now with the introduction of the aforementioned `tseslint.config()` function.

Related discussion: https://github.com/typescript-eslint/typescript-eslint/discussions/6090 (I didn't know whether to comment there, or start a new thread, I'm happy to close this and move the discussion there)

### Before you submit your discussion, please confirm the following. If any of these required steps are not taken, we may not be able to review your RFC. Help us to help you!

- [X] I have [read the discussions guidelines](https://typescript-eslint.io/contributing/discussions) and per those guidelines, this fits a discussion category, _not_ a help request or standard issue.
- [X] I have [searched for related discussions](https://github.com/typescript-eslint/typescript-eslint/discussions) and [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues) and found none that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

**Comment by Shinigami92** [comment]
Hi 🥰
Always happy to join forces from my side.
Right now I still do not have knowledge about v9's FlatConfig, but @sxzz did add support for that.

**Comment by bradzacher** [comment]
TL;DR - I think that migrating off of `eslint-define-config` onto the canonical solution of `typescript-eslint` is the best bet for flat configs.

---

The big reason that I didn't want us to get into hosting and owning the `eslint-define-config` solution back when it was proposed was because there wasn't really a good solution for typed plugins in legacy configs beyond using declaration merging. I was certainly bullish on the idea - type-safe configs are an amazing concept - but I didn't necessarily think that pushing the ecosystem to define types in the legacy config world was the best of things as it put a lot of burden on the user (eg having to add triple slash refs for each plugin that supported it). I also sadly just didn't have the time to pitch in and contribute to expanding the project.

With flat configs we now have one major thing that we didn't have before: all plugins and configs are imported into the config file. This means that there's no longer a triple-slash-ref requirement and we can extract types directly from the package (or perhaps an `@types/` package for the plugin!!!).

So with this we can document a solution and [release tooling](https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/rule-schema-to-typescript-types) that allows a plugin to generate a type signature as a build step that we can then extract (using the `typescript-eslint.config` function) to provide types in the config.

That was actually a big part of the reason that I chose to release the `typescript-eslint.config` function to begin with -- because I saw it as a way for us to finally have an "in" in the config so that we can provide strict types / autocomplete in JS files (without a JSDoc annotation or an extension or anything) and it provides us a platform to build better types (eg plugin rule names, plugin rule options, plugin settings) and also to build new features (eg the first release came with an `extends` property cos we felt it was needed).

**Comment by Shinigami92** [comment]
Yesterday I did my first steps with eslint v9 & FlatConfig

Using the [`tseslint.config`](https://github.com/faker-js/faker/pull/2798/files#diff-a32a0887ed9d1d707bbb3b845b7df7fd40e673c47e7b60a3ebd896b68d3b8839R9) function, I'm tempted to fully deprecate eslint-define-config in the long run and we could think about providing _augmented global properties_ for the `Rules` here:

https://github.com/typescript-eslint/typescript-eslint/blob/074310029694f92323d4d91f0fd31df39698d0e7/packages/utils/src/ts-eslint/Config.ts#L166

So `@eslint-types/*` can than somehow ship Rule types and append them to the `Rules` type that can than be consumed by `typescript-eslint` instead of `eslint-define-config`.

What do you think about this @bradzacher?

**Comment by Shinigami92** [comment]
Just as a note, I deprecated `eslint-define-config` in favor of eslint's native `defineConfig`
@Lehoczky you can close this discussion if possible for you
